### PR TITLE
Enhanced parallelism, much cleaner code (dev-master-9)

### PR DESCRIPTION
**Pull request by only-a-ptr**
Highlights:
* GUI improvements with DearIMGUI (top menubar, new skeletonview) showcased here: https://github.com/RigsOfRods/rigs-of-rods/pull/1333
* Tidy-up of physics data structures -> thinner, more sensible
* Project 'AsyncScene' -> major multithreading improvement, updating visual scene (flexbodies, flares, props+animations, cab, dashboard....) now runs in parallel with physics

### CODE REVIEW

**Comment by tritonas00** [comment]
> /home/babis/Downloads/rigsofrods-next-build-db-1.8/rigsofrods/src/rigs-of-rods/source/main/gfx/DustManager.cpp: In member function ‘void RoR::GfxScene::InitScene(Ogre::SceneManager*)’:
/home/babis/Downloads/rigsofrods-next-build-db-1.8/rigsofrods/src/rigs-of-rods/source/main/gfx/DustManager.cpp:68:22: error: ‘make_unique’ is not a member of ‘std’
         m_heathaze = std::make_unique<HeatHaze>();

**Comment by tritonas00** [comment]
It doesn't work

**Comment by tritonas00** [comment]
2018-08-23-130010_1920x1080_scrot

:stuck_out_tongue_winking_eye:

**Comment by AnotherFoxGuy** [comment]
@only-a-ptr For gcc you need to add ```set (CMAKE_CXX_STANDARD 14)``` to the main cmakelist.txt, see: https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html

**Comment by tritonas00** [comment]
Apart from some wheels issues, everything seem to work fine

**Comment by only-a-ptr** [comment]
>@tritonas00 Apart from some wheel issues, everything seem to work fine

I must have broken the wheels just yesterday, it always worked before

Anyway considering the scale of changes, there must be more glitches somewhere, look harder 😉

**Comment by tritonas00** [comment]
I'm diggin it

**Comment by tritonas00** [comment]
Another one:

Load simple terrain and spawn agora. Go to top bar see Actors, it shows agora fine.
Remove agora and respawn it. Go to top bar see Actors: empty

**Comment by only-a-ptr** [comment]
I fixed the wheels and added the terrain fault tolerance fix. More to come.

**Comment by tritonas00** [comment]
hmm
> /home/babis/Downloads/rigsofrods-next-build-db-1.8/rigsofrods/src/rigs-of-rods/source/main/gameplay/RoRFrameListener.cpp: In member function ‘bool SimController::CameraManagerMouseMoved(const OIS::MouseEvent&)’:
/home/babis/Downloads/rigsofrods-next-build-db-1.8/rigsofrods/src/rigs-of-rods/source/main/gameplay/RoRFrameListener.cpp:2351:44: error: ‘bool RoR::CameraManager::mouseMoved(const OIS::MouseEvent&)’ is protected within this context
     return m_camera_manager.mouseMoved(_arg);
                                            ^
In file included from /home/babis/Downloads/rigsofrods-next-build-db-1.8/rigsofrods/src/rigs-of-rods/source/main/gameplay/RoRFrameListener.h:25:0,
                 from /home/babis/Downloads/rigsofrods-next-build-db-1.8/rigsofrods/src/rigs-of-rods/source/main/gameplay/RoRFrameListener.cpp:21:
/home/babis/Downloads/rigsofrods-next-build-db-1.8/rigsofrods/src/rigs-of-rods/source/main/gfx/camera/CameraManager.h:148:10: note: declared protected here
     bool mouseMoved(const OIS::MouseEvent& _arg);

**Comment by only-a-ptr** [comment]
@tritonas00 Why does Visual Studio always compile broken code? 🤢

**Comment by tritonas00** [comment]
Update:

gavril reset crash from https://github.com/RigsOfRods/rigs-of-rods/pull/1538#issuecomment-415392614
happens on master also

**Comment by tritonas00** [comment]
this https://github.com/RigsOfRods/rigs-of-rods/issues/1437 is still happening

**Comment by tritonas00** [comment]
Another issue: 

Load Auriga map. Once loaded try to walk/move the ror bot. It doesn't go anywhere. Master works.

**Comment by tritonas00** [comment]
I gathered all issues here: https://github.com/RigsOfRods/rigs-of-rods/pull/1538#issuecomment-415392614

**Comment by ulteq** [comment]
Almost twice as much FPS for me with 20 trucks in the scene! I'm loving it.

Master: 23 FPS
This: 40 FPS

**Comment by tritonas00** [comment]
This PR probably fixes also this: https://github.com/RigsOfRods/rigs-of-rods/issues/1356 because now i can do:

2018-09-01-210914_1366x768_scrot

Nice!

**Comment by tritonas00** [comment]
@only-a-ptr please remove https://github.com/RigsOfRods/rigs-of-rods/pull/1538/commits/322bf6fa5f9393b5071947774312cb705bf669b9 from this PR, it doesn't work and you already stated that is broken: https://github.com/RigsOfRods/rigs-of-rods/pull/1460#issuecomment-365777452

**Comment by only-a-ptr** [comment]
> @ulteq Almost twice as many FPS for me with 20 trucks in the scene! I'm loving it.

Heh, and it wasn't even motivated by performance, but rather code clarity. All sim first, all gfx later. The `bool doUpdate` mess in N/B processing was too much :sunglasses:

**Comment by only-a-ptr** [comment]
I definitely resolved the N-Labs crash, see https://github.com/RigsOfRods/rigs-of-rods/issues/1437#issuecomment-417950544

**Comment by tritonas00** [comment]
Ok at least it shows the error message. But there are other maps also like simple test terrain 12 cars that crash because of this. This https://github.com/RigsOfRods/rigs-of-rods/pull/1549 fixes the issue without modifying the maps. Please include that hot fix also. 

Why break something that worked? :confused:

**Comment by only-a-ptr** [comment]
> @tritonas00 But there are other maps also...

In that case I'll have to take care of it, but know that it's not me who breaks things, it's the modders who create faulty terrains that only work by chance. The documentation is clear on this. @Michael10055 's _empty pagefile_ "fix" or @ulteq 's _1549_ "fix" are just hacks which create terrain page with absolutely bogus default parameters. :nauseated_face:

**Comment by tritonas00** [comment]
fair enough

**Comment by ulteq** [comment]
> with absolutely bogus default parameters.

Can't we provide proper default parameters?

**Comment by only-a-ptr** [comment]
@ulteq Default parameters for terrain? Like 'if there's no heightmap, use [Mount Fuji](https://en.wikipedia.org/wiki/Mount_Fuji)?' 😆

It's our code, we can simply disable the OgreTerrain altogether if the modder doesn't want it. My point is, why think of workarounds when we can create sensible designs?

**Comment by ulteq** [comment]
All we have to do to make the code backwards compatible is to remove those two lines:
https://github.com/only-a-ptr/rigs-of-rods/blob/cccce32df16e0463dbd7688cb565f9ef6cf833a0/source/main/terrain/TerrainGeometryManager.cpp#L331-L332
We can leave a warning in if you like.

**Comment by ulteq** [comment]
Where did those go: https://github.com/RigsOfRods/rigs-of-rods/pull/1538/commits/fe4277795e848b3fe9aa8c903e51da71b8b4b6c2#diff-abbec6ed672585ca1ea0964858716deeL3783

Is it not needed anymore?

**Comment by only-a-ptr** [comment]
@ulteq /unrelated/ it went here: https://github.com/RigsOfRods/rigs-of-rods/commit/fe4277795e848b3fe9aa8c903e51da71b8b4b6c2#diff-0b090e76f9de202df6134e778c74b554R2336

> All we would have to do, to make the code backwards compatible, is to remove this line...
Sure, go ahead, just add comment `// Ignoring error for backwards compatibility`

**Comment by ulteq** [comment]
More bugs:

- Ties are invisble
- The interstate 8st trailer  spawns at (0,0,0)

**Comment by tritonas00** [comment]
> But only if there really are multiple maps with this problem. @tritonas00, are there?

Only two i know.. @Michael10055 would help here...

File: source/main/gfx/GfxActor.cpp [hunk 1]
```diff
@@ -1202,3 +1333,992 @@ void RoR::GfxActor::CalculateDriverPos(Ogre::Vector3& out_pos, Ogre::Quaternion&
...
+    for (prop_t& prop: m_props)
+    {
+        if (!prop.scene_node)
+            continue;
+
+        // -- quick ugly port from `Actor::NotifyCameraChanged()`~ 06/2018
+        const bool mo_visible = (prop.cameramode == -2 || prop.cameramode == m_simbuf.simbuf_cur_cinecam);
+        if (prop.mo != nullptr)
+        {
+            prop.mo->setVisible(mo_visible);
```
Comment (ulteq) [inline source/main/gfx/GfxActor.cpp]: This used to be `setMeshEnabled`.

File: source/main/gfx/GfxActor.cpp [hunk 2]
```diff
@@ -1202,3 +1333,992 @@ void RoR::GfxActor::CalculateDriverPos(Ogre::Vector3& out_pos, Ogre::Quaternion&
...
+    }
+
+    NodeData* nodes = this->GetSimNodeBuffer();
+
+    for (prop_t& prop: m_props)
+    {
+        if (!prop.scene_node)
+            continue;
+
+        // -- quick ugly port from `Actor::NotifyCameraChanged()`~ 06/2018
```
Comment (ulteq) [inline source/main/gfx/GfxActor.cpp]: What if the prop is a `scene_node`?

Comment (only-a-ptr) [reply source/main/gfx/GfxActor.cpp]: > What if the prop is a scene_node?

I don't understand the question.

The `if (prop.scene_node == nullptr)` comes from the original code, somebody needs to fix it - no scenenode means setup failed and the prop should never make it to `m_props`.

File: source/main/gfx/GfxActor.cpp (hunk 1 above)
Comment (only-a-ptr) [reply source/main/gfx/GfxActor.cpp]: Perhaps a mistake, I really don't know, the whole `MeshObject` is a piece of ****, see this branch I just made: https://github.com/only-a-ptr/rigs-of-rods/commits/MeshObject-cleanup

File: source/main/gfx/GfxActor.cpp (hunk 2 above)
Comment (ulteq) [reply source/main/gfx/GfxActor.cpp]: I think the `// -- quick ugly port from `Actor::NotifyCameraChanged()`~ 06/2018` needs to always run.

Comment (only-a-ptr) [reply source/main/gfx/GfxActor.cpp]: Eh, that's just one of many things which need to be researched and cleaned up. I work in small iterations, first I cut+paste code with all quirks, then test, then I revisit it later. And everything related to the `AsyncScene` rewrite is in the 'pasted with quirks' stage.

Comment (ulteq) [reply source/main/gfx/GfxActor.cpp]: Yeah, but those subtle code changes are gonna bite us in the ass later on. :smile:

Comment (only-a-ptr) [reply source/main/gfx/GfxActor.cpp]: That's a topic for a long forum thread, bugs will always happen, but short commits help pinpointing them, even if it means messier code in the meantime (that's what comments are for). I wouldn't stress causing a few visual glitches, what troubles me in the long run are physics/collisions/scripting.

**Comment by ulteq** [comment]
This does not work with 'Load' actors, which stay `SLEEPING` after being spawned: https://github.com/RigsOfRods/rigs-of-rods/blob/6839b19dd519e6881149c0acea7669ab08809545/source/main/physics/BeamFactory.cpp#L462-L476
Example trucks:
- https://forum.rigsofrods.org/downloads.php?do=file&id=170
- https://forum.rigsofrods.org/downloads.php?do=file&id=402

**Comment by only-a-ptr** [comment]
@ulteq Good spot, the `isPreloadedWithTerrain` should really be `/*is anyhow spawned as SLEEPING*/`. 

I'm trying to get rid of all duplicate/redundant updates in the spawn code. Historically we had a single program flow for any sort of spawn (with map, from script, user action...) which needed to handle all specifics. I'm trying to split it apart.

**Comment by tritonas00** [comment]
Another issue:

Spawn skyvan press F2 to open the back the door. Press F1. It stops. Press F1 again. It still goes down. When it goes full down press F1, it goes up.

Master works at is should: Pressing F2 or F21 instantly opens or closes the door.

**Comment by ulteq** [comment]
`UpdateRods()` needs to account for `bm_disabled` and `bm_broken`.

**Comment by only-a-ptr** [comment]
> @ulteq I think UpdateRods() needs to account for bm_disabled and bm_broken.

Precisely, I just postponed with this note: https://github.com/RigsOfRods/rigs-of-rods/commit/27835c8f08695d567b7a47d1ce69ed29086c21d7#diff-0b090e76f9de202df6134e778c74b554R786

**Comment by ulteq** [comment]
> Precisely, I just postponed it with this note: https://github.com/RigsOfRods/rigs-of-rods/commit/27835c8f08695d567b7a47d1ce69ed29086c21d7#diff-0b090e76f9de202df6134e778c74b554R786

All right, then I'll leave that to you.

**Comment by only-a-ptr** [comment]
What's left to fix/revert here? All the bugs listed above are fixed.

I'm currently working on spawn logic, but the changes are getting big and it's not exactly critical to do right now: https://github.com/only-a-ptr/rigs-of-rods/commits/devmaster9-spawn-codepath

**Pull request merged**